pallets/file-system/src/utils.rs

```diff
@@ -578,96 +625,381 @@ where
+                }
 
-        // Get the current root of the bucket where the file will be stored.
-        let bucket_root = expect_or_err!(
-            <T::Providers as shp_traits::ReadBucketsInterface>::get_root_bucket(&bucket_id),
-            "Failed to get root for bucket, when it was already checked to exist",
-            Error::<T>::BucketNotFound
-        );
+                if !failed_file_keys.is_empty() {
+                    for rejected_file_key in failed_file_keys {
+                        failed
+                            .try_push(rejected_file_key)
+                            .map_err(|_| Error::<T>::TooManyStorageRequestResponses)?;
+                    }
+                }
+            }
+
+            if let Some(rejected_file_keys) = file_key_responses.reject {
+                let mut rejected: BoundedVec<
+                    (MerkleHash<T>, RejectedStorageRequestReason),
+                    MaxBatchMspRespondStorageRequests<T>,
+                > = BoundedVec::default();
+
+                for file_key in rejected_file_keys.iter() {
+                    let storage_request_metadata = match <StorageRequests<T>>::get(&file_key.0) {
+                        Some(metadata) => metadata,
+                        None => {
+                            failed
+                                .try_push((file_key.0, Error::<T>::StorageRequestNotFound.into()))
+                                .map_err(|_| Error::<T>::TooManyStorageRequestResponses)?;
+                            continue;
+                        }
+                    };
+
+                    if let Err(e) = Self::cleanup_storage_request(
+                        EitherAccountIdOrProviderId::ProviderId(msp_id),
+                        file_key.0,
+                        &storage_request_metadata,
+                    ) {
+                        failed
+                            .try_push((file_key.0, e))
+                            .map_err(|_| Error::<T>::TooManyStorageRequestResponses)?;
+                        continue;
+                    }
+
+                    rejected
+                        .try_push((file_key.0, file_key.1.clone()))
+                        .map_err(|_| Error::<T>::TooManyStorageRequestResponses)?;
+                }
+
+                if !rejected.is_empty() {
+                    let rejected_batch = MspRejectedBatchStorageRequests {
+                        file_keys: rejected,
+                        bucket_id: bucket_id.clone(),
+                        owner: owner.clone(),
+                    };
+
+                    batch_responses
+                        .try_push(BatchResponses::Rejected(rejected_batch))
+                        .map_err(|_| Error::<T>::TooManyBatchResponses)?;
+                }
+            }
+
+            if !failed.is_empty() {
+                let failed_batch = MspFailedBatchStorageRequests {
+                    file_keys: failed,
+                    bucket_id: bucket_id.clone(),
```

Comment:
Same here^

Reply:
I think both Provider/Bucket IDs and Hashes (file keys, roots, etc) implement copy, so no need to clone them